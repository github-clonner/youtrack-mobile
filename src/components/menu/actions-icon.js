/* @flow */

import React, {PureComponent} from 'react';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

type Props = {};

export default class ActionsIcon extends PureComponent<Props, void> {

  render() {
    return (
      <Icon name="dots-horizontal" size={26}/>
    );
  }
}
