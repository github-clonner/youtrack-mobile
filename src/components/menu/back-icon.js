/* @flow */

import React, {PureComponent} from 'react';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import {Platform} from 'react-native';

type Props = {
  color?: string
};

export default class BackIcon extends PureComponent<Props, void> {

  render() {
    const isAndroid = Platform.OS === 'android';
    return (
      <Icon
        name={isAndroid ? 'arrow-left' : 'chevron-left'}
        size={isAndroid ? 23 : 30}
        color={this.props.color}
      />
    );
  }
}
