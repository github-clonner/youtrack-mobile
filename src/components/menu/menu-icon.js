/* @flow */

import React, {PureComponent} from 'react';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

type Props = {};

export default class MenuIcon extends PureComponent<Props, void> {

  render() {
    return (
      <Icon name="menu" size={24}/>
    );
  }
}
