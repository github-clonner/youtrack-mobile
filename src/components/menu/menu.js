/* @flow */
import {View, ScrollView, Text, TouchableWithoutFeedback, TouchableOpacity, Linking} from 'react-native';
import React, {Component} from 'react';
import styles from './menu.styles';
import {VERSION_STRING} from '../usage/usage';
import Drawer from 'react-native-drawer';
import Router from '../router/router';
import Auth from '../auth/auth';
import MenuItem from './menu__item';
import clicksToShowCounter from '../debug-view/clicks-to-show-counter';
import {connect} from 'react-redux';
import {
  openMenu,
  closeMenu,
  openFeaturesView,
} from '../../actions/app-actions';

import type {AgileUserProfile} from '../../flow/Agile';
import Feature from '../feature/feature';
import ConnectedAccounts from './menu__connected-accounts';

const CURRENT_YEAR = (new Date()).getFullYear();
const MENU_WIDTH = 280;

function openPrivacyPolicy() {
  Linking.openURL('https://www.jetbrains.com/company/privacy.html');
}


type Props = {
  children?: React$Element<any>,
  show: boolean,
  auth: Auth,
  issueQuery: ?string,
  agileProfile: AgileUserProfile,
  onOpen: () => any,
  onClose: () => any,
  openFeaturesView: () => any
};

type DefaultProps = {
  onOpen: () => any,
  onClose: () => any
};

export class Menu extends Component<Props, void> {
  static defaultProps: DefaultProps = {
    onOpen: () => {},
    onClose: () => {}
  };

  _openIssueList = () => {
    this.props.onClose();
    Router.IssueList();
  };

  _openAgileBoard = () => {
    this.props.onClose();
    Router.AgileBoard();
  };

  _openInbox = () => {
    this.props.onClose();
    Router.Inbox();
  };

  _getSelectedAgileBoard = () => {
    const {agileProfile} = this.props;

    if (!agileProfile || !agileProfile.defaultAgile) {
      return '';
    }

    const lastAgileName = agileProfile.defaultAgile?.name || '';
    let lastSprintName;

    if (agileProfile.visitedSprints) {
      lastSprintName = agileProfile.visitedSprints.find(
        sprint => sprint.agile && (sprint.agile.id === agileProfile.defaultAgile?.id)
      )?.name || '';
    }

    return [lastAgileName, lastSprintName].filter(Boolean).join(', ');
  };

  _renderMenu() {
    const {auth, issueQuery, openFeaturesView} = this.props;
    if (!auth) { //TODO: menu renders right after logOut by some reason.
      return null;
    }

    return (
      <ScrollView
        testID="menuContainer"
        style={styles.scrollContainer}>
        <View style={styles.menuContainer}>
          <View style={styles.accounts}><ConnectedAccounts/></View>

          <View style={styles.menuItems}>
            <MenuItem
              label={'Issues'}
              description={issueQuery || 'No query'}
              onPress={this._openIssueList}
            />

            <MenuItem
              label={'Agile Boards'}
              testId="pageAgileBoards"
              description={this._getSelectedAgileBoard()}
              onPress={this._openAgileBoard}
            />

            <Feature version={'2018.3'}>
              <MenuItem
                label={'Notifications'}
                description={''}
                onPress={this._openInbox}
              />
            </Feature>
          </View>

          <View style={styles.menuFooter}>
            <TouchableWithoutFeedback onPress={() => clicksToShowCounter(openFeaturesView, 'open features list')}>
              <Text style={styles.footerText}>YouTrack Mobile {VERSION_STRING}</Text>
            </TouchableWithoutFeedback>

            <View style={styles.spacer}/>
            <Text style={styles.footerText}>© 2000—{CURRENT_YEAR} JetBrains</Text>
            <Text style={styles.footerText}>All rights reserved</Text>

            <View style={styles.spacer}/>
            <TouchableOpacity style={styles.buttonLink} onPress={openPrivacyPolicy}>
              <Text style={styles.linkText}>Privacy Policy</Text>
            </TouchableOpacity>
          </View>
        </View>
      </ScrollView>
    );
  }

  render() {
    const {children, show, onOpen, onClose} = this.props;

    return (
      <Drawer
        testID="menuDrawer"
        type="static"
        open={show}
        content={this._renderMenu()}
        tapToClose={true}
        onOpen={onOpen}
        onClose={onClose}
        openDrawerOffset={viewport => viewport.width - MENU_WIDTH}
        captureGestures={true}
        panOpenMask={12}
      >
        {children}
      </Drawer>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    show: state.app.showMenu,
    auth: state.app.auth,
    issueQuery: state.issueList.query,
    agileProfile: state.agile.profile,
    ...ownProps
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onOpen: () => dispatch(openMenu()),
    onClose: () => dispatch(closeMenu()),
    openFeaturesView: () => dispatch(openFeaturesView()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Menu);

