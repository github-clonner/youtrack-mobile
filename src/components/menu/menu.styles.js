import {StyleSheet} from 'react-native';
import {COLOR_PINK, UNIT, COLOR_GRAY, COLOR_FONT} from '../../components/variables/variables';

export default StyleSheet.create({
  accounts: {
    borderColor: COLOR_FONT,
    borderBottomWidth: 0.5
  },
  scrollContainer: {
    backgroundColor: 'black'
  },
  menuContainer: {
    flex: 1,
    flexDirection: 'column',
    paddingTop: UNIT * 1.5
  },
  menuItems: {
    flexGrow: 2,
    padding: UNIT * 2,
    paddingTop: UNIT,
  },
  menuFooter: {
    padding: UNIT * 2,
    paddingTop: UNIT * 4
  },
  spacer: {
    marginTop: UNIT
  },
  footerText: {
    fontSize: 12,
    lineHeight: UNIT * 2,
    color: COLOR_GRAY
  },
  buttonLink: {
    padding: UNIT,
    paddingLeft: 0
  },
  linkText: {
    color: COLOR_PINK,
    fontSize: 12
  }
});
