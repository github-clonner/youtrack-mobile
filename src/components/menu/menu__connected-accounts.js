/* @flow */

import React, {Component} from 'react';
import {connect} from 'react-redux';

import {
  addAccount,
  changeAccount, closeMenu,
  openDebugView,
  removeAccountOrLogOut
} from '../../actions/app-actions';

import type {StorageState} from '../storage/storage';
import Accounts from '../account/accounts';

type DefaultProps = {
  onClose: () => void
};

type Props = {
  otherAccounts: Array<StorageState>,
  isChangingAccount: ?boolean,

  onClose: () => any,
  onLogOut: () => any,
  onAddAccount: () => any,
  onChangeAccount: (account: StorageState) => any,

  openDebugView: () => any
};


export class ConnectedAccounts extends Component<Props, void> {
  static defaultProps: DefaultProps = {
    onClose: () => {}
  };

  render() {
    const {
      openDebugView,
      isChangingAccount,
      onAddAccount,
      onChangeAccount,
      otherAccounts,
      onLogOut,
      onClose
    } = this.props;

    return (
      <Accounts
        isChangingAccount={isChangingAccount}
        openDebugView={openDebugView}
        otherAccounts={otherAccounts}
        onLogOut={onLogOut}
        onAddAccount={onAddAccount}
        onChangeAccount={onChangeAccount}
        onClose={onClose}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    otherAccounts: state.app.otherAccounts,
    isChangingAccount: state.app.isChangingAccount,
    ...ownProps
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onClose: () => dispatch(closeMenu()),
    onLogOut: () => dispatch(removeAccountOrLogOut()),
    onAddAccount: () => dispatch(addAccount()),
    onChangeAccount: (account: StorageState) => dispatch(changeAccount(account)),
    openDebugView: () => dispatch(openDebugView()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Accounts);

