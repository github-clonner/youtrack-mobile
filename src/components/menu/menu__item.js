/* @flow */

import {TouchableOpacity, View, Text, Image} from 'react-native';
import React, {PureComponent} from 'react';

import {next} from '../icon/icon';

import styles from './menu__item.styles';

import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';

type Props = {
  label: string,
  description?: string,
  onPress: () => any,
  testId?: string,
  style?: ViewStyleProp
}

export default class MenuItem extends PureComponent<Props, void> {

  render() {
    const {onPress, style = null, testId = null, label, description = ''} = this.props;

    return (
      <TouchableOpacity
        testID={testId}
        style={[styles.menuItem, style]}
        activeOpacity={0.4}
        onPress={onPress}
      >
        <View style={styles.menuItemTopLine}>
          <Text style={styles.menuItemText}>{label}</Text>
          <Image style={styles.menuItemIcon} source={next}/>
        </View>
        <Text style={styles.menuItemSubtext}>{description}</Text>
      </TouchableOpacity>
    );
  }
}
