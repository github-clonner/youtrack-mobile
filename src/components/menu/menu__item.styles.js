import {StyleSheet} from 'react-native';
import {
  UNIT,
  COLOR_GRAY,
  COLOR_FONT_ON_BLACK, COLOR_FONT
} from '../../components/variables/variables';

const bottomBorder = {
  borderColor: COLOR_FONT,
  borderBottomWidth: 0.5
};

export default StyleSheet.create({
  menuItem: {
    paddingTop: UNIT / 2,
    paddingBottom: UNIT * 2,
    marginBottom: UNIT,
    ...bottomBorder
  },
  menuItemTopLine: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  menuItemIcon: {
    height: UNIT * 2,
    resizeMode: 'contain'
  },
  menuItemText: {
    lineHeight: 18,
    fontSize: 17,
    fontWeight: 'bold',
    color: COLOR_FONT_ON_BLACK
  },
  menuItemSubtext: {
    lineHeight: 19,
    paddingTop: UNIT / 2,
    color: COLOR_GRAY,
    fontSize: 14
  }
});
