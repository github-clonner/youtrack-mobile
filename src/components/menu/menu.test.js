import React from 'react';
import {Text} from 'react-native';

import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

import {Menu} from './menu';
import Router from '../router/router';


describe('<Menu/>', () => {

  const rootTestID = 'menuDrawer';
  const containerTestID = 'menuContainer';
  let wrapper;
  let instance;

  describe('Render', () => {

    it('should match a snapshot', () => {
      render({auth: {}, agileUserProfile: {}});

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render component', () => {
      render({auth: {}, agileUserProfile: {}});

      expect(findByTestId(rootTestID)).toHaveLength(1);
      expect(findChildByTestId(rootTestID, containerTestID)).toHaveLength(1);
    });

    it('should not render menu container if `auth` is not provided', () => {
      render({});

      expect(findByTestId(rootTestID)).toHaveLength(1);
      expect(findChildByTestId(rootTestID, containerTestID)).toHaveLength(0);
    });

    it('should render menu children', () => {
      const menuChildrenTestId = 'menuChildren';
      render({auth: {}, children: <Text testID={menuChildrenTestId}>node</Text>});

      expect(findByTestId(rootTestID)).toHaveLength(1);
      expect(findChildByTestId(rootTestID, menuChildrenTestId)).toHaveLength(1);
    });
  });


  describe('_getSelectedAgileBoard', () => {
    const idMock = 'id';
    const agileNameMock = 'agileName';
    const lastVisitedSprintNameMock = 'sprintName';

    it('should return empty string if an `agileProfile` is not provided', () => {
      render({auth: {}});

      expect(instance._getSelectedAgileBoard()).toEqual('');
    });

    it('should return empty string if there is no `defaultAgile` filed in an `agileProfile`', () => {
      renderWithAuth({});

      expect(instance._getSelectedAgileBoard()).toEqual('');
    });

    it('should return empty string if a `defaultAgile` filed has no name', () => {
      renderWithAuth({
        defaultAgile: {}
      });

      expect(instance._getSelectedAgileBoard()).toEqual('');
    });

    it('should return empty string if a `defaultAgile` is empty but `visitedSprints` is not empty', () => {
      renderWithAuth({
        defaultAgile: {},
        visitedSprints: [
          {agile: {id: 'id'}}
        ]
      });

      expect(instance._getSelectedAgileBoard()).toEqual('');
    });

    it('should return agile name', () => {
      renderWithAuth({
        defaultAgile: {name: agileNameMock},
        visitedSprints: []
      });

      expect(instance._getSelectedAgileBoard()).toEqual(agileNameMock);
    });

    it('should return agile and last visited sprint', () => {

      renderWithAuth({
        defaultAgile: {name: agileNameMock, id: idMock},
        visitedSprints: [{
          name: lastVisitedSprintNameMock,
          agile: {id: idMock}
        }]
      });

      expect(instance._getSelectedAgileBoard()).toEqual(`${agileNameMock}, ${lastVisitedSprintNameMock}`);
    });

  });


  describe('', () => {
    let _methods;

    beforeEach(() => {
      renderWithAuth();
      _methods = [Router.IssueList, Router.AgileBoard];
      Router.IssueList = jest.fn();
      Router.AgileBoard = jest.fn();

      wrapper.setProps({
        onClose: jest.fn()
      });
    });

    afterEach(() => {
      Router.IssueList = _methods[0];
      Router.AgileBoard = _methods[1];
    });

    describe('_openIssueList', () => {
      it('should close menu', () => {
        instance._openIssueList();

        expect(wrapper.props().onClose).toHaveBeenCalled();
      });

      it('should redirect to Issue list', () => {
        instance._openIssueList();

        expect(Router.IssueList).toHaveBeenCalled();
      });
    });

    describe('_openAgileBoard', () => {
      it('should close menu', () => {
        instance._openAgileBoard();

        expect(wrapper.props().onClose).toHaveBeenCalled();
      });

      it('should redirect to Agile board', () => {
        instance._openIssueList();

        expect(Router.IssueList).toHaveBeenCalled();
      });
    });
  });


  function render({auth, agileUserProfile, children}) {
    wrapper = doShallow(auth, agileUserProfile, children);
    instance = wrapper.instance();
  }

  function renderWithAuth(agileUserProfile) {
    return render({auth: {}, agileUserProfile});
  }

  function findByTestId(testId) {
    return wrapper && wrapper.find({testID: testId});
  }

  function findChildByTestId(parentTestId, targetTestId) {
    const parentShallowWrapper = findByTestId(parentTestId);
    if (parentShallowWrapper) {
      return parentShallowWrapper.dive().find({testID: targetTestId});
    }
  }

  function doShallow(auth, agileUserProfile, children) {
    return shallow(
      <Menu
        auth={auth}
        show={true}
        onOpen={() => {}}
        onClose={() => {}}
        openFeaturesView={() => {}}
        agileProfile={agileUserProfile}
        issueQuery={''}
      >{children}</Menu>
    );
  }
});
