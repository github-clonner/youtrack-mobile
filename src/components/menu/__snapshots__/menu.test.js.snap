// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu/> Render should match a snapshot 1`] = `
<Drawer
  acceptDoubleTap={false}
  acceptPan={true}
  acceptPanOnDrawer={true}
  acceptTap={false}
  captureGestures={true}
  closedDrawerOffset={0}
  content={
    <ScrollViewMock
      style={
        Object {
          "backgroundColor": "black",
        }
      }
      testID="menuContainer"
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
            "paddingTop": 12,
          }
        }
      >
        <View
          style={
            Object {
              "borderBottomWidth": 0.5,
              "borderColor": "#444444",
            }
          }
        >
          <Connect(Accounts) />
        </View>
        <View
          style={
            Object {
              "flexGrow": 2,
              "padding": 16,
              "paddingTop": 8,
            }
          }
        >
          <MenuItem
            description="No query"
            label="Issues"
            onPress={[Function]}
          />
          <MenuItem
            description=""
            label="Agile Boards"
            onPress={[Function]}
            testId="pageAgileBoards"
          />
          <Connect(Feature)
            version="2018.3"
          >
            <MenuItem
              description=""
              label="Notifications"
              onPress={[Function]}
            />
          </Connect(Feature)>
        </View>
        <View
          style={
            Object {
              "padding": 16,
              "paddingTop": 32,
            }
          }
        >
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <Text
              style={
                Object {
                  "color": "#CCC",
                  "fontSize": 12,
                  "lineHeight": 16,
                }
              }
            >
              YouTrack Mobile 
              0.1 (build 0)
            </Text>
          </TouchableWithoutFeedback>
          <View
            style={
              Object {
                "marginTop": 8,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "#CCC",
                "fontSize": 12,
                "lineHeight": 16,
              }
            }
          >
            © 2000—
            2020
             JetBrains
          </Text>
          <Text
            style={
              Object {
                "color": "#CCC",
                "fontSize": 12,
                "lineHeight": 16,
              }
            }
          >
            All rights reserved
          </Text>
          <View
            style={
              Object {
                "marginTop": 8,
              }
            }
          />
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            style={
              Object {
                "padding": 8,
                "paddingLeft": 0,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FE0082",
                  "fontSize": 12,
                }
              }
            >
              Privacy Policy
            </Text>
          </TouchableOpacity>
        </View>
      </View>
    </ScrollViewMock>
  }
  disabled={false}
  elevation={0}
  initializeOpen={false}
  negotiatePan={false}
  onClose={[Function]}
  onOpen={[Function]}
  open={true}
  openDrawerOffset={[Function]}
  panCloseMask={null}
  panOpenMask={12}
  panThreshold={0.25}
  side="left"
  styles={Object {}}
  tapToClose={true}
  testID="menuDrawer"
  tweenDuration={250}
  tweenEasing="linear"
  tweenHandler={null}
  type="static"
  useInteractionManager={false}
/>
`;
